packages/financial_analysis/categorize.py

```diff
@@ -322,7 +326,32 @@ def categorize_expenses(
     if not isinstance(page_size, int) or page_size <= 0:
         raise ValueError("page_size must be a positive integer")
 
-    # Determine pages lazily via _paginate(); no need to precompute total count.
+    # Group before LLM: normalize merchant/description; pick one exemplar per group
+    def _norm_merchant_key(tx: Mapping[str, Any]) -> str | None:
+        raw = tx.get("merchant") or tx.get("description")
+        if raw is None:
+            return None
+        s = unicodedata.normalize("NFKC", str(raw)).strip()
+        if not s:
+            return None
+        return " ".join(s.split()).casefold()
+
+    by_key: dict[str, list[int]] = {}
+    singleton_indices: list[int] = []
+    for i, tx in enumerate(original_seq):
+        k = _norm_merchant_key(tx)
+        if k is None:
+            singleton_indices.append(i)
+        else:
+            by_key.setdefault(k, []).append(i)
+    # Deterministic exemplar per group: the smallest absolute index
+    exemplars: list[int] = []
+    for idxs in by_key.values():
+        exemplars.append(min(idxs))
+    exemplars.extend(singleton_indices)  # singletons act as their own group
+    exemplars.sort()
+
+    n_groups = len(exemplars)
```

Comment:
@CharlieHelps move the grouping code into a helper function that's clearly named